### PR TITLE
Bug/RMSRCBUILD-196 release branch is not pushed to remote

### PR DESCRIPTION
**Pull request by Frazerus**
[Jira Issue 196](https://re-motion.atlassian.net/browse/RMSRCBUILD-196)

### CODE REVIEW

File: ReleaseProcessAutomation/Steps/PipelineSteps/PushNewReleaseBranch.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+using ReleaseProcessAutomation.Git;
+
+namespace ReleaseProcessAutomation.Steps.PipelineSteps;
+
+public interface IPushNewReleaseBranch
+{
+  void Execute (string releaseBranchName, string firstAncestorBranchName);
+}
+
+public class PushNewReleaseBranch
```
Comment (a-ctor) [inline ReleaseProcessAutomation/Steps/PipelineSteps/PushNewReleaseBranch.cs:12]: This class and its interface should have a `Step` suffix, otherwise they are located in the wrong folder

File: ReleaseProcessAutomation/Steps/PipelineSteps/PushNewReleaseBranch.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    _gitClient = gitClient;
+    _config = config;
+  }
+  
+  public void Execute (string releaseBranchName, string firstAncestorBranchName)
+  {
+    var remoteNames = _config.RemoteRepositories.RemoteNames;
+    _gitClient.PushToRepos(remoteNames, releaseBranchName);
+    if(!string.IsNullOrEmpty(firstAncestorBranchName))
+      _gitClient.PushToRepos(remoteNames, firstAncestorBranchName);
```
Comment (a-ctor) [inline ReleaseProcessAutomation/Steps/PipelineSteps/PushNewReleaseBranch.cs:29]: It is kind of strange that we pass in two branch names and do the same thing (push) but the second one is optional. We should remove one of the parameters and let the caller decide how often they want to call us. The `ReleaseOnMasterStep` would call us twice while the `ReleasePatchStep` just once.

File: ReleaseProcessAutomation/Steps/PipelineSteps/ReleasePatchStep.cs [hunk 3]
```diff
@@ -50,34 +51,37 @@ public class ReleasePatchStep : ReleaseProcessStepBase, IReleasePatchStep
...
+      if (!currentBranchName.StartsWith("master"))
       {
         const string message = "Cannot release a patch for master when not on master branch.";
         throw new InvalidOperationException(message);
       }
     }
     else
     {
-      if (!GitClient.IsOnBranch("hotfix/"))
+      if (!currentBranchName.StartsWith("hotfix/"))
```
Comment (a-ctor) [inline ReleaseProcessAutomation/Steps/PipelineSteps/ReleasePatchStep.cs:84]: Do we need those changes? They seem irrelevant to the issue at hand.

File: ReleaseProcessAutomation.Tests/Steps/Releases/ReleaseNonPreReleaseFromDevelopTests.cs [hunk 4]
```diff
@@ -116,22 +120,22 @@ public void Execute_NothingThrows_ShouldCallNextStep()
...
-
-
+              gitClientStub.Object,
+              readInputStub.Object,
+              _continueReleaseOnMasterMock.Object,
+              _pushNewReleaseBranchMock.Object,
+              _config,
+              _msBuildInvokerMock.Object,
+              _consoleMock.Object);
+
```
Comment (a-ctor) [inline ReleaseProcessAutomation.Tests/Steps/Releases/ReleaseNonPreReleaseFromDevelopTests.cs:130]: Formatting was correct.

**Review by a-ctor — CHANGES_REQUESTED** [review]
The branch name for this PR seems incorrect. There is also the branch `bugfix/RMSRCBUILD-196-release-branch-is-not-pushed-to-remote` which has the correct naming. We should use the branch with the correct name and delete the other branch. Be aware that this PR is closed when the branch is deleted so we should do this after all the reviews are done.

Also there are a few formatting issue left (e.g. `(_=>...` instead of `(_ => ...`). Please look through the changes and fix those formatting issues.

File: ReleaseProcessAutomation/Steps/PipelineSteps/PushNewReleaseBranch.cs (hunk 2 above)
Comment (Frazerus) [reply ReleaseProcessAutomation/Steps/PipelineSteps/PushNewReleaseBranch.cs:29]: ReleasePatchStep would also need to call this twice, I think the only thing we could/should do is to remove the optionality of the second argument.

File: ReleaseProcessAutomation/Steps/PipelineSteps/ReleasePatchStep.cs (hunk 3 above)
Comment (Frazerus) [reply ReleaseProcessAutomation/Steps/PipelineSteps/ReleasePatchStep.cs:84]: Well we need the current branch name already, so why not use it and save some git calls? I know this doesnt fit this issue, perfectly, it does come in contact with it however. 

Can still revert it if that is not enough of an argument ;)

Comment (a-ctor) [reply ReleaseProcessAutomation/Steps/PipelineSteps/ReleasePatchStep.cs:84]: `GitClient.IsOnBranch("bla")` and `currentBranch.StartsWith("bla")` do not seem like they test the same condition. We can keep the changes but we should address that issue.

File: ReleaseProcessAutomation/Steps/PipelineSteps/PushNewReleaseBranch.cs (hunk 2 above)
Comment (a-ctor) [reply ReleaseProcessAutomation/Steps/PipelineSteps/PushNewReleaseBranch.cs]: If the second parameter is no longer optional then there is no reason why the method has to push two branches at the same time. In either case I think it is better to have the functionality of "pushing a branch" rather then "pushing a branch and optionally another one as well". I think this is better because the user can decide what they want to do. Otherwise, if you just want to push a single branch you either have to duplicate the logic or pass an empty string just for the sake of providing values.

File: ReleaseProcessAutomation/Steps/PipelineSteps/ReleasePatchStep.cs (hunk 3 above)
Comment (Frazerus) [reply ReleaseProcessAutomation/Steps/PipelineSteps/ReleasePatchStep.cs:88]: Not quite sure how to address that issue, naming? Wrapper method? Git branches are defined by names in this implementation, so I think it should seem like startsWith checks the same, but IDK.

File: ReleaseProcessAutomation.UnitTests/Steps/Releases/ReleaseNonPreReleaseFromDevelopTests.cs [hunk 5]
```diff
@@ -51,6 +52,7 @@ public void Setup ()
     _continueReleaseOnMasterMock = new Mock<IContinueReleaseOnMasterStep>();
     _consoleMock = new Mock<IAnsiConsole>();
     _jiraFunctionalityMock = new Mock<IJiraFunctionality>();
+    _pushNewReleaseBranchMock = new Mock<IPushNewReleaseBranchStep>();
```
Comment (a-ctor) [inline ReleaseProcessAutomation.UnitTests/Steps/Releases/ReleaseNonPreReleaseFromDevelopTests.cs:55]: I just realized that using a loose mock here is kind of bad. If the step is called when we don't want it to be called then such a test would not fail. We should use a strict mock and explicitly setup when we expect the call.

File: ReleaseProcessAutomation.UnitTests/Steps/Releases/ReleasePatchStepTests.cs [hunk 6]
```diff
@@ -54,6 +55,7 @@ public void Setup ()
     _contineReleasePatchMock = new Mock<IContinueReleasePatchStep>();
     _consoleStub = new Mock<IAnsiConsole>();
     _jiraFunctionalityMock = new Mock<IJiraFunctionality>();
+    _pushNewReleaseBranchMock = new Mock<IPushNewReleaseBranchStep>();
```
Comment (a-ctor) [inline ReleaseProcessAutomation.UnitTests/Steps/Releases/ReleasePatchStepTests.cs:58]: See previous comment -> strict mock

File: ReleaseProcessAutomation/Steps/PipelineSteps/ReleasePatchStep.cs [hunk 7]
```diff
@@ -52,36 +53,39 @@ public class ReleasePatchStep : ReleaseProcessStepBase, IReleasePatchStep
...
+    if (currentBranchName == null)
     {
       const string message = "Could not get the name of the current branch.";
       throw new InvalidOperationException(message);
     }
 
     if (onMaster)
     {
-      if (!GitClient.IsOnBranch("master"))
+      if (!currentBranchName.Equals("master"))
```
Comment (a-ctor) [inline ReleaseProcessAutomation/Steps/PipelineSteps/ReleasePatchStep.cs:80]: In C# the preferred way of checking equality is using `==`. Compared to Java when `.Equals` and `==` do different kinds of equality checks, `==` and `.Equals` are equivalent in C#. They both compare the value of the string and not the reference. `.Equals` is useful in situations where you want special string comparisons though, for example, case-insensitive string comparision, as you can supply a second argument for the type of comparison.

File: ReleaseProcessAutomation/Steps/PipelineSteps/ReleasePatchStep.cs [hunk 8]
```diff
@@ -52,36 +53,39 @@ public class ReleasePatchStep : ReleaseProcessStepBase, IReleasePatchStep
...
       {
         const string message = "Cannot release a patch for master when not on master branch.";
         throw new InvalidOperationException(message);
       }
     }
     else
     {
-      if (!GitClient.IsOnBranch("hotfix/"))
+      if (!currentBranchName.StartsWith("hotfix/"))
```
Comment (a-ctor) [inline ReleaseProcessAutomation/Steps/PipelineSteps/ReleasePatchStep.cs:88]: This should also be changed to `==`

**Review by a-ctor — CHANGES_REQUESTED** [review]

File: ReleaseProcessAutomation/Steps/PipelineSteps/ReleasePatchStep.cs (hunk 7 above)
Comment (Frazerus) [reply ReleaseProcessAutomation/Steps/PipelineSteps/ReleasePatchStep.cs:80]: Ok, thank you for the input :)

File: ReleaseProcessAutomation/Steps/PipelineSteps/ReleasePatchStep.cs (hunk 8 above)
Comment (Frazerus) [reply ReleaseProcessAutomation/Steps/PipelineSteps/ReleasePatchStep.cs:88]: Im not sure youre correct here, we dont know which version number will be appended to the hotfix/*** so we cant really check for equality

Comment (a-ctor) [reply ReleaseProcessAutomation/Steps/PipelineSteps/ReleasePatchStep.cs:88]: I should learn to read :D

File: ReleaseProcessAutomation.UnitTests/Steps/Releases/ReleasePatchStepTests.cs [hunk 9]
```diff
@@ -118,29 +122,34 @@ public void Execute_OnHotfixWithoutErrors_CallsNextStep ()
...
     var nextVersion = new SemanticVersion { Patch = 1 };
     var nextJiraVersion = new SemanticVersion();
     _gitClientStub.Setup(_ => _.IsWorkingDirectoryClean()).Returns(true);
     _gitClientStub.Setup(_ => _.IsCommitHash("")).Returns(true);
     _gitClientStub.Setup(_ => _.IsOnBranch("hotfix/")).Returns(true);
-    _gitClientStub.Setup(_ => _.GetCurrentBranchName()).Returns("release/v0.0.1");
+    _gitClientStub.Setup(_ => _.GetCurrentBranchName()).Returns("hotfix/v0.0.1");
     var nextPossibleVersions = nextVersion.GetNextPossibleVersionsHotfix();
     _inputReaderStub.Setup(_ => _.ReadVersionChoice(It.IsAny<string>(), nextPossibleVersions)).Returns(nextJiraVersion);
+    _pushNewReleaseBranchMock.Setup(_=>_.Execute("release/v0.0.1", "hotfix/v0.0.1" ));
```
Comment (a-ctor) [inline ReleaseProcessAutomation.UnitTests/Steps/Releases/ReleasePatchStepTests.cs:135]:
```suggestion
    _pushNewReleaseBranchMock.Setup(_ => _.Execute("release/v0.0.1", "hotfix/v0.0.1" ));
```

File: ReleaseProcessAutomation.UnitTests/Steps/Releases/ReleasePatchStepTests.cs [hunk 10]
```diff
@@ -118,29 +122,34 @@ public void Execute_OnHotfixWithoutErrors_CallsNextStep ()
...
         _contineReleasePatchMock.Object,
+        _pushNewReleaseBranchMock.Object,
         _consoleStub.Object,
         _jiraFunctionalityMock.Object);
 
     Assert.That(
         () => patchStep.Execute(nextVersion, "", true, false, false, false),
         Throws.Nothing);
+    
+    _pushNewReleaseBranchMock.Verify(_=>_.Execute("release/v0.0.1", "hotfix/v0.0.1" ));
```
Comment (a-ctor) [inline ReleaseProcessAutomation.UnitTests/Steps/Releases/ReleasePatchStepTests.cs:152]:
```suggestion
    _pushNewReleaseBranchMock.Verify(_ => _.Execute("release/v0.0.1", "hotfix/v0.0.1" ));
```

**Review by a-ctor — APPROVED** [review]
Two formatting remarks that would be nice if we fix them before merging. Also please squash everything into a single commit  `RMSRCBUILD-196 xxxx` before merging.
